stackslib/src/net/api/gettenure.rs

```diff
@@ -0,0 +1,324 @@
+use crate::net::api::getblock_v3::NakamotoBlockStream;
+use crate::net::http::{
+    parse_bytes, Error, HttpBadRequest, HttpChunkGenerator, HttpContentType, HttpNotFound,
+    HttpRequest, HttpRequestContents, HttpRequestPreamble, HttpResponse, HttpResponseContents,
+    HttpResponsePayload, HttpResponsePreamble, HttpServerError, HttpVersion,
+};
+use crate::net::httpcore::{
+    HttpRequestContentsExtensions, RPCRequestHandler, StacksHttp, StacksHttpRequest,
+    StacksHttpResponse,
+};
+use crate::net::{Error as NetError, StacksNodeState, TipRequest, MAX_HEADERS};
+use crate::util_lib::db::{DBConn, Error as DBError};
+
+#[derive(Clone)]
+pub struct RPCNakamotoTenureRequestHandler {
+    /// Block to start streaming from. It and its ancestors will be incrementally streamed until one of
+    /// hte following happens:
+    /// * we reach the first block in the tenure
+    /// * we would exceed MAX_MESSAGE_LEN bytes transmitted if we started sending the next block
+    pub block_id: Option<StacksBlockId>,
+}
+
+impl RPCNakamotoTenureRequestHandler {
+    pub fn new() -> Self {
+        Self { block_id: None }
+    }
+}
+
+pub struct NakamotoTenureStream {
+    /// stream for the current block
+    pub block_stream: NakamotoBlockStream,
+    /// connection to the headers DB
+    pub headers_conn: DBConn,
+    /// total bytess sent so far
+    pub total_sent: u64,
+}
+
+impl NakamotoTenureStream {
+    pub fn new(
+        chainstate: &StacksChainState,
+        block_id: StacksBlockId,
+        consensus_hash: ConsensusHash,
+        parent_block_id: StacksBlockId,
+    ) -> Result<Self, ChainError> {
+        let block_stream =
+            NakamotoBlockStream::new(chainstate, block_id, consensus_hash, parent_block_id)?;
+        let headers_conn = chainstate.reopen_db()?;
+        Ok(NakamotoTenureStream {
+            block_stream,
+            headers_conn,
+            total_sent: 0,
+        })
+    }
+
+    /// Start streaming the next block (i.e. the parent of the block we last streamed).
+    /// Return Ok(true) if we can fit the block into the stream.
+    /// Return Ok(false) if not. The caller will need to call this RPC method again with the block
+    /// ID of the last block it received.
+    /// Return Err(..) on DB error
+    pub fn next_block(&mut self) -> Result<bool, ChainError> {
+        let parent_header = NakamotoChainState::get_block_header_nakamoto(
+            &self.headers_conn,
+            &self.block_stream.parent_block_id,
+        )?
+        .ok_or(ChainError::NoSuchBlockError)?;
+
+        // stop sending if the parent is an epoch2 block
+        let Some(parent_nakamoto_header) = parent_header.anchored_header.as_stacks_nakamoto()
+        else {
+            return Ok(false);
+        };
+
+        // stop sending if the parent is in a different tenure
+        if parent_nakamoto_header.consensus_hash != self.block_stream.consensus_hash {
+            return Ok(false);
```

Comment:
Should this and the above condition be errors? This function returns `Ok(false)` whether it's out of space (and the caller should try again), or the block is invalid

Reply:
@jcnelson to confirm, but I think that once `MAX_MESSAGE_LEN`, a new request must be made, so the caller actually shouldn't try again to get more chunks from the stream (i.e., the HTTP stream should close).

Reply:
`Ok(false)` here means "end of stream", not an error.  @kantai that is correct -- the caller would resume downloading the tenure as of the last received block (whose `block_id` is the path parameter), and use the `stop=` query parameter to indicate to the server to stop sending blocks once a particular block is reached (such as, for example, if the caller is querying only new unconfirmed tenure blocks).  This is handled by the `NakamotoTenureDownloader` and `NakamotoUnconfirmedTenureDownloader` state machines in the downloader PR -- they'll both remain in their `GetTenureBlocks(last_block_id)` state until the full tenure has been fetched, and will update `last_block_id` if the blocks received did not include the tenure-start block.

The `stop=` query parameter isn't present in this PR; it's part of the downloader itself (which modified this API).